core/modules/mysql/MySqlConnection.cc

```diff
@@ -161,24 +149,41 @@ MySqlConnection::selectDb(std::string const& dbName) {
 // MySqlConnection
 // private:
 ////////////////////////////////////////////////////////////////////////
-bool
-MySqlConnection::_initMySql() {
-    _isConnected = false; // reset.
-    _mysql = NULL;
-    std::lock_guard<std::mutex> g(_mysqlShared);
-    if(!_mysqlReady) {
-        int rc = mysql_library_init(0, NULL, NULL);
-        assert(0 == rc);
-        _mysqlReady = true;
-    }
-    return true;
-}
 
 MYSQL* MySqlConnection::_connectHelper() {
+    // We must call mysql_library_init() exactly once before calling mysql_init
+    // because it is not thread safe. Both mysql_library_init and mysql_init
+    // call mysql_thread_init, and so we must arrange to call mysql_thread_end
+    // when the calling thread exists. We do this by allocating a thread
+    // local object that calls mysql_thread_end from its destructor.
+    //
+    // C++11 thread_local storage cannot be used until the baseline compiler is
+    // GCC 4.8, so we must fall back to boost. A significant caveat from the
+    // boost documentation is that "on some platforms, cleanup of thread-
+    // specific data is not performed for threads created with the platform's
+    // native API". Fortunately, even though we are creating threads via
+    // the STL and not boost, cleanup seems to be performed on POSIX systems.
+    // However, this likely does not work on Windows.
+    static std::once_flag initialized;
+    static boost::thread_specific_ptr<MySqlThreadJanitor> janitor;
+
+    struct InitializeMysqlLibrary {
+        void operator()() {
+            int rc = mysql_library_init(0, NULL, NULL);
+            assert(0 == rc);
+            if (!janitor.get()) {
+                janitor.reset(new MySqlThreadJanitor);
+            }
```

Comment:
We wouldn't really expect janitor.get() != NULL, because of the call_once below, right?  Maybe this should be an assert instead of an if?